rls/src/main/java/io/grpc/rls/internal/LbPolicyConfiguration.java

```diff
@@ -0,0 +1,369 @@
+    private ConnectivityStateInfo connectivityStateInfo =
+        ConnectivityStateInfo.forNonError(ConnectivityState.IDLE);
+    private SubchannelPicker picker;
+
+    public ChildPolicyWrapper(
+        String target,
+        ChildLoadBalancerHelperProvider childLbHelperProvider,
+        @Nullable  ChildLbStatusListener childLbStatusListener) {
+      this.target = target;
+      this.helper =
+          new ChildPolicyReportingHelper(childLbHelperProvider, this, childLbStatusListener);
+    }
+
+    String getTarget() {
+      return target;
+    }
+
+    void setPicker(SubchannelPicker picker) {
+      this.picker = checkNotNull(picker, "picker");
+    }
+
+    SubchannelPicker getPicker() {
+      return picker;
+    }
+
+    ChildPolicyReportingHelper getHelper() {
+      return helper;
+    }
+
+    void setConnectivityStateInfo(ConnectivityStateInfo connectivityStateInfo) {
+      this.connectivityStateInfo = connectivityStateInfo;
+    }
+
+    ConnectivityStateInfo getConnectivityStateInfo() {
+      return connectivityStateInfo;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      ChildPolicyWrapper that = (ChildPolicyWrapper) o;
+      return Objects.equals(target, that.target)
+          && Objects.equals(helper, that.helper)
+          && Objects.equals(connectivityStateInfo, that.connectivityStateInfo)
+          && Objects.equals(picker, that.picker);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(target, helper, connectivityStateInfo, picker);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("target", target)
+          .add("helper", helper)
+          .add("connectivityStateInfo", connectivityStateInfo)
+          .add("picker", picker)
+          .toString();
+    }
+  }
+
+  private static final class RefCountedChildPolicyWrapper
+      implements ObjectPool<ChildPolicyWrapper> {
+
+    private final AtomicLong refCnt = new AtomicLong(1);
+    @Nullable
+    private ChildPolicyWrapper childPolicyWrapper;
+
+    private RefCountedChildPolicyWrapper(ChildPolicyWrapper childPolicyWrapper) {
+      this.childPolicyWrapper = checkNotNull(childPolicyWrapper, "childPolicyWrapper");
+    }
+
+    @Override
+    public ChildPolicyWrapper getObject() {
+      long curr = refCnt.getAndIncrement();
+      if (curr <= 0) {
+        throw new IllegalStateException("ChildPolicyWrapper is already released");
+      }
+      return childPolicyWrapper;
+    }
+
+    @Override
+    @Nullable
+    public ChildPolicyWrapper returnObject(Object object) {
+      checkState(
+          childPolicyWrapper == object,
+          "returned object doesn't match the pooled childPolicyWrapper");
+      long newCnt = refCnt.decrementAndGet();
+      if (newCnt == 0) {
```

Comment:
What's the expected usecase here? `getObject()` once and `returnObject()` twice()?

Reply:
this handles when the object is just released, free up the reference.

Reply:
If you call x times `getObject()`, you need call x+1 times of `returnObject()` to get here, is that the contract?

Reply:
`create` itself is counted as `getObject()` since the object is passed in.

Reply:
> create itself is counted as getObject() 
But you can not use the object unless you call `getObject()` explicitly.  The API contract looks strange.  If you just need to you the object once, you have to explicitly call  `getObject()` once, and you need `returnObject()` twice to get here.

Reply:
why it cannot use the object before getObject()? anyways, okay it depends on what is the focus. focused more on the "RefCounted" it makes more sense to have it as 1, if you focused more on "ObjectPool" part starting with 0 makes more sense. refCount = 0 is released state (ref counted bytebuf starts with 1 for example), but for the sake of consistency i can follow the other convention, also it implements ObjectPool makes me leaning toward that.

Reply:
I see, it is used directly for the first time without calling `getObject()`
```java
RefCountedChildPolicyWrapper wrapper = RefCountedChildPolicyWrapper.of(childPolicyWrapper);
childPolicyMap.put(target, wrapper);
return childPolicyWrapper;
```